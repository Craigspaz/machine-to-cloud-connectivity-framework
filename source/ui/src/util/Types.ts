// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import React from 'react';

/**
 * The Amplify configuration input type.
 * The configuration will be generated while launching the solution.
 * @interface AmplifyConfigurationInput
 */
export interface AmplifyConfigurationInput {
  apiEndpoint: string;
  identityPoolId: string;
  loggingLevel: 'ERROR' | 'WARN' | 'INFO' | 'DEBUG' | 'VERBOSE';
  region: string;
  userPoolId: string;
  webClientId: string;
}

/**
 * The Amplify configuration response type for the web application
 * @interface AmplifyConfigurationResponse
 */
export interface AmplifyConfigurationResponse {
  API: {
    endpoints: { name: string, endpoint: string, region: string }[];
  };
  Auth: {
    identityPoolId: string;
    region: string;
    userPoolId: string;
    userPoolWebClientId: string;
  };
}

/**
 * The connection definition
 * @interface ConnectionDefinition
 */
export interface ConnectionDefinition {
  control: ConnectionControl;
  connectionName: string;
  protocol: MachineProtocol;
  area?: string;
  machineName?: string;
  opcDa?: OpcDaDefinition;
  opcUa?: OpcUaDefinition;
  process?: string;
  sendDataToIoTSitewise?: boolean;
  sendDataToIoTTopic?: boolean;
  sendDataToKinesisDataStreams?: boolean;
  siteName?: string;
}

/**
 * Common connection definition
 * @interface CommonDefinition
 */
export interface CommonDefinition {
  machineIp: string;
  serverName: string;
}

/**
 * OPC DA connection definition
 * @interface OpcDaDefinition
 * @extends CommonDefinition
 */
export interface OpcDaDefinition extends CommonDefinition {
  interval: number | string;
  iterations: number | string;
  listTags?: string[];
  tags?: string[];
}

/**
 * OPC UA connection definition
 * @interface OpcUaDefinition
 * @extends CommonDefinition
 */
export interface OpcUaDefinition extends CommonDefinition {
  port?: number | string;
}

/**
 * @enum The connection control types
 */
export enum ConnectionControl {
  DEPLOY = 'deploy',
  START = 'start',
  STOP = 'stop',
  UPDATE = 'update',
  DELETE = 'delete',
  PUSH = 'push',
  PULL = 'pull'
}

/**
 * The list connections item type
 * @interface ListConnectionsItem
 */
export interface ListConnectionsItem {
  connectionName: string;
  protocol: MachineProtocol;
  status: ConnectionControl;
  sendDataToIoTSitewise: boolean;
  sendDataToIoTTopic: boolean;
  sendDataToKinesisDataStreams: boolean;
  machineName?: string;
}

/**
 * The list connections response type
 * @interface ListConnectionsResponse
 */
export interface ListConnectionsResponse {
  connections: ListConnectionsItem[];
  nextToken?: string;
}

/**
 * The control connection response type.
 * @interface ControlConnectionResponse
 */
export interface ControlConnectionResponse {
  connectionName: string;
  control: ConnectionControl;
  result: string;
}

/**
 * The create and update connection response type
 * @interface CreateUpdateConnectionResponse
 */
export interface CreateUpdateConnectionResponse {
  connectionName: string;
  control: ConnectionControl.DEPLOY | ConnectionControl.UPDATE;
  message: string;
}

/**
 * The default modal properties
 * @interface ModalProps
 */
interface ModalProps {
  show: boolean;
  hide: Function;
}

/**
 * The message modal properties
 * @interface MessageModalProps
 * @extends ModalProps
 */
export interface MessageModalProps extends ModalProps {
  message: string | React.ReactNode;
  modalType: MessageModalType;
  confirmAction?: Function;
}

/**
 * @enum The message modal types
 */
export enum MessageModalType {
  MESSAGE,
  CONFIRM
}

/**
 * The connection modal properties
 * @interface ConnectionModalProps
 * @extends ModalProps
 */
export interface ConnectionModalProps extends ModalProps {
  initConnection: GetConnectionResponse;
  isUpdating: boolean;
  setIsUpdating: Function;
  connectionName?: string;
}

/**
 * A connection response
 * @interface GetConnectionResponse
 */
export interface GetConnectionResponse {
  connectionName: string;
  protocol: MachineProtocol;
  sendDataToIoTSitewise: boolean;
  sendDataToIoTTopic: boolean;
  sendDataToKinesisDataStreams: boolean;
  area?: string;
  machineName?: string;
  opcDa?: OpcDaDefinition;
  opcUa?: OpcUaDefinition;
  process?: string;
  siteName?: string;
  listTags?: string;
  tags?: string;
}

/**
 * @enum The connection protocols
 */
export enum MachineProtocol {
  OPCDA = 'opcda',
  OPCUA = 'opcua'
}

/**
 * The form properties for the each protocols.
 * @interface FormProps
 */
export interface FormProps {
  connection: GetConnectionResponse;
  onChange: Function;
  errors: KeyStringValue;
}

/**
 * The form properties for the OPC UA protocol.
 * @interface OpcUaFormProps
 * @extends FormProps
 */
export interface OpcUaFormProps extends FormProps {
  region: string;
}

/**
 * @type The HTML form control element type
 */
export type FormControlElement = HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement;

/**
 * The general key, value JSON object type which allows undefined value
 * @interface KeyStringValue
 */
export interface KeyStringValue {
  [key: string]: string | undefined;
}

/**
 * The machine logs modal properties
 * @interface ConnectionLogsModalProps
 * @extends ModalProps
 */
export interface ConnectionLogsModalProps extends ModalProps {
  connectionName: string;
}

/**
 * @enum Log type
 */
export enum LogType {
  INFO = 'info',
  ERROR = 'error'
}

/**
 * The list logs item type
 * @interface ListLogsItem
 */
export interface ListLogsItem {
  connectionName: string;
  timestamp: number;
  logType: LogType;
  message: string;
}

/**
 * The list logs response type
 * @interface ListLogsResponse
 */
export interface ListLogsResponse {
  logs: ListLogsItem[];
  nextToken?: string;
}

/**
 * The alphanumeric characters, hyphens, and underscores validation type
 * @interface AlphaNumericValidationProps
 */
export interface AlphaNumericValidationProps {
  value: string;
  maxLength: number;
  errors: KeyStringValue;
  errorKeyName: string;
  errorMessage: string;
}

/**
 * @enum The pagination type
 */
export enum PaginationType {
  PREV,
  NEXT
}