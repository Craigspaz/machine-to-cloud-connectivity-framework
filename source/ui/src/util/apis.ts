// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { API } from '@aws-amplify/api';
import { ApiProps } from './types';
import { API_NAME } from './utils';

/**
 * Requests API to the backend.
 * @param params The API parameters
 * @returns The response from the API call.
 */
export async function requestApi(params: ApiProps): Promise<unknown> {
  const { method, path, options } = params;
  return API[method](API_NAME, path, { ...options });
}
