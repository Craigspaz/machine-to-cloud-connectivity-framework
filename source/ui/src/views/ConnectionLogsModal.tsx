// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { API } from '@aws-amplify/api';
import { I18n, Logger } from '@aws-amplify/core';
import React, { useCallback, useEffect, useState } from 'react';
import Button from 'react-bootstrap/Button';
import Alert from 'react-bootstrap/Alert';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import Jumbotron from 'react-bootstrap/Jumbotron';
import Modal from 'react-bootstrap/Modal';
import ProgressBar from 'react-bootstrap/ProgressBar';
import Row from 'react-bootstrap/Row';
import Table from 'react-bootstrap/Table';
import MessageModal from './MessageModal';
import { ConnectionLogsModalProps, ListLogsItem, ListLogsResponse, MessageModalType, PaginationType } from '../util/Types';
import { API_NAME, getErrorMessage } from '../util/Utils';

const logger = new Logger('ConnectionLogModal');

/**
 * Renders the connection logs modal.
 * @param props The connection logs modal properties
 * @returns The connection logs modal
 */
export default function ConnectionLogsModal(props: ConnectionLogsModalProps) {
  const { show, hide, connectionName } = props;
  const [logs, setLogs] = useState<ListLogsItem[]>([]);
  const [loading, setLoading] = useState<boolean>(false);
  const [showMessagemMessageModal, setShowMessageMessageModal] = useState<boolean>(false);
  const [messageModalMessage, setMessageModalMessage] = useState<React.ReactNode>('');
  const [pageIndex, setPageIndex] = useState<number>(0);
  const [pageToken, setPageToken] = useState<string[]>(['']);

  /**
   * Gets the connection logs.
   * @param paginationType The pagination type to get the connection logs
   */
  const getLogs = useCallback(async (paginationType?: PaginationType) => {
    setLoading(true);

    let nextToken: string;

    switch (paginationType) {
      case PaginationType.PREV:
        nextToken = pageToken[pageIndex - 1];
        break;
      case PaginationType.NEXT:
        nextToken = pageToken[pageIndex + 1];
        break;
      default:
        nextToken = '';
        break;
    }

    try {
      const encodedConnectionName = encodeURIComponent(connectionName);
      const response: ListLogsResponse = await API.get(API_NAME, `/logs/${encodedConnectionName}`, {
        queryStringParameters: { nextToken: nextToken && nextToken !== '' ? nextToken : undefined }
      });

      switch (paginationType) {
        case PaginationType.PREV:
          setPageIndex(prevIndex => prevIndex - 1);
          break;
        case PaginationType.NEXT:
          /**
           * Due to inconsistency, it can't compare with the next index item.
           * Therefore, it checks if the page token array has the next token or not.
           */
          if (response.nextToken && !pageToken.includes(response.nextToken)) {
            setPageToken([
              ...pageToken,
              response.nextToken
            ]);
          }
          setPageIndex(prevIndex => prevIndex + 1);
          break;
        default:
          setPageIndex(0);
          setPageToken(response.nextToken ? ['', response.nextToken] : ['']);
          break;
      }

      setLogs(response.logs);
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      logger.error(errorMessage);

      setShowMessageMessageModal(true);
      setMessageModalMessage(
        <Alert variant="danger">
          {I18n.get('error.message.get.logs')}
          <br />
          {I18n.get('error')}: <code>{JSON.stringify(errorMessage)}</code>
        </Alert>
      )
    }

    setLoading(false);
  }, [connectionName, pageToken, pageIndex]);

  /**
   * React useEffect hook.
   */
  useEffect(() => {
    if (connectionName) getLogs();
  }, [connectionName]); // eslint-disable-line react-hooks/exhaustive-deps

  /**
   * Renders the message of the empty logs.
   * @returns Empty logs component
   */
  function EmptyLogs(): JSX.Element {
    return (
      <Jumbotron className="text-align-center" id="empty-logs-jumbotron">
        <p className="empty-p">{I18n.get('info.message.no.logs')}</p>
      </Jumbotron>
    );
  }

  return (
    <>
      <Modal show={show} onHide={() => hide()} id="connection-log-modal" animation={false} size="xl">
        <Modal.Header closeButton>
          <strong>{I18n.get('connection.logs')}</strong>: {connectionName}
        </Modal.Header>
        <Modal.Body>
          {loading && <ProgressBar animated now={100} />}
          {!loading && logs.length === 0 && <EmptyLogs />}
          {!loading && logs.length > 0 &&
            <>
              <Table size="md" id="connection-logs-table">
                <thead>
                  <tr>
                    <th>{I18n.get('message')}</th>
                    <th>{I18n.get('log.type')}</th>
                    <th>{I18n.get('date')}</th>
                  </tr>
                </thead>
                <tbody>
                  {
                    logs.map((log: ListLogsItem, index: number) => {
                      return (
                        <tr key={`logs-${log.connectionName}-${index}`}>
                          <td>
                            <Form>
                              <Form.Group>
                                <Form.Control as="textarea" rows={5} readOnly value={log.message} />
                              </Form.Group>
                            </Form>
                          </td>
                          <td className="uppercase-text">{log.logType}</td>
                          <td>{new Date(log.timestamp).toISOString()}</td>
                        </tr>
                      );
                    })
                  }
                </tbody>
              </Table>
              <Row>
                <Col>
                  <Button id="prev-logs-page-button" size="sm" className="uppercase-text" disabled={pageIndex === 0}
                    onClick={() => getLogs(PaginationType.PREV)}><i className="bi bi-chevron-double-left" /> {I18n.get('prev.page')}</Button>
                </Col>
                <Col className="justify-content-end grid">
                  <Button id="next-logs-page-button" size="sm" className="uppercase-text" disabled={pageToken.length === pageIndex + 1}
                    onClick={() => getLogs(PaginationType.NEXT)}>{I18n.get('next.page')} <i className="bi bi-chevron-double-right" /></Button>
                </Col>
              </Row>
            </>
          }
        </Modal.Body>
        <Modal.Footer>
          <Button id="connection-logs-modal-close-button" onClick={() => hide()} className="uppercase-text" size="sm">{I18n.get('close')}</Button>
        </Modal.Footer>
      </Modal>
      <MessageModal show={showMessagemMessageModal}
        hide={() => setShowMessageMessageModal(false)}
        message={messageModalMessage}
        modalType={MessageModalType.MESSAGE}
      />
    </>
  );
}