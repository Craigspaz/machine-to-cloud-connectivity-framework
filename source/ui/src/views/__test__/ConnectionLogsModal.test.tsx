// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { API } from '@aws-amplify/api';
import { I18n } from '@aws-amplify/core';
import { queryByAttribute, render, screen, waitFor } from '@testing-library/react';
import { ConnectionLogsModalProps, ListLogsResponse, LogType } from '../../util/Types';
import ConnectionLogsModal from '../ConnectionLogsModal';

// Mock API
const mockAPI = {
  get: jest.fn()
};
jest.mock('@aws-amplify/api');
API.get = mockAPI.get;

const getById = queryByAttribute.bind(null, 'id');

beforeEach(() => {
  mockAPI.get.mockReset();
});

test('renders the connection message modal component when connection logs are empty', async () => {
  const response: ListLogsResponse = { logs: [] };
  mockAPI.get.mockResolvedValueOnce(response);
  const props: ConnectionLogsModalProps = {
    show: true,
    hide: () => console.log('hide'),
    connectionName: 'mock-connection-id'
  };
  const connectionMessageModal = render(<ConnectionLogsModal {...props} />);

  await waitFor(() => {
    getById(connectionMessageModal.container, 'empty-logs-jumbotron');
  });
  expect(mockAPI.get).toHaveBeenCalledTimes(1);
  expect(mockAPI.get).toHaveBeenCalledWith('M2C2Api', '/logs/mock-connection-id', { queryStringParameters: { nextToken: undefined } });
  expect(connectionMessageModal.baseElement).toMatchSnapshot();
});

test('renders the connection message modal component when getting connection logs fails', async () => {
  mockAPI.get.mockRejectedValueOnce('Error');
  const props: ConnectionLogsModalProps = {
    show: true,
    hide: () => console.log('hide'),
    connectionName: 'mock-connection-id'
  };
  const connectionMessageModal = render(<ConnectionLogsModal {...props} />);

  await waitFor(() => {
    screen.findByText(I18n.get('error.message.get.logs'));
  });
  expect(mockAPI.get).toHaveBeenCalledTimes(1);
  expect(mockAPI.get).toHaveBeenCalledWith('M2C2Api', '/logs/mock-connection-id', { queryStringParameters: { nextToken: undefined } });
  expect(connectionMessageModal.baseElement).toMatchSnapshot();
});

test('renders the connection message modal component when there are connection logs', async () => {
  const response: ListLogsResponse = {
    logs: [
      { connectionName: 'mock-connection-id', logType: LogType.INFO, timestamp: new Date(0).getTime(), message: JSON.stringify({ message: 'info message' }) },
      { connectionName: 'mock-connection-id', logType: LogType.ERROR, timestamp: new Date(0).getTime(), message: JSON.stringify({ message: 'error message' }) },
    ]
  };
  mockAPI.get.mockResolvedValueOnce(response);
  const props: ConnectionLogsModalProps = {
    show: true,
    hide: () => console.log('hide'),
    connectionName: 'mock-connection-id'
  };
  const connectionMessageModal = render(<ConnectionLogsModal {...props} />);

  await waitFor(() => {
    getById(connectionMessageModal.container, 'connection-logs-table');
  });
  expect(mockAPI.get).toHaveBeenCalledTimes(1);
  expect(mockAPI.get).toHaveBeenCalledWith('M2C2Api', '/logs/mock-connection-id', { queryStringParameters: { nextToken: undefined } });
  expect(connectionMessageModal.baseElement).toMatchSnapshot();
});

test('renders the connection message modal component when there are connection logs and the next token', async () => {
  const response: ListLogsResponse = {
    logs: [
      { connectionName: 'mock-connection-id', logType: LogType.INFO, timestamp: new Date(0).getTime(), message: JSON.stringify({ message: 'info message' }) },
      { connectionName: 'mock-connection-id', logType: LogType.ERROR, timestamp: new Date(0).getTime(), message: JSON.stringify({ message: 'error message' }) },
    ],
    nextToken: JSON.stringify({ connectionName: 'mock-connection-id', timestamp: new Date(0).getTime() })
  };
  mockAPI.get.mockResolvedValueOnce(response);
  const props: ConnectionLogsModalProps = {
    show: true,
    hide: () => console.log('hide'),
    connectionName: 'mock-connection-id'
  };
  const connectionMessageModal = render(<ConnectionLogsModal {...props} />);

  await waitFor(() => {
    getById(connectionMessageModal.container, 'connection-logs-table');
  });
  expect(mockAPI.get).toHaveBeenCalledTimes(1);
  expect(mockAPI.get).toHaveBeenNthCalledWith(1, 'M2C2Api', '/logs/mock-connection-id', { queryStringParameters: { nextToken: undefined } });
  expect(connectionMessageModal.baseElement).toMatchSnapshot();
});