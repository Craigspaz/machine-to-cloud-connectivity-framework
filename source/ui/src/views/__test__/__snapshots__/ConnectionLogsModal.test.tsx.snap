// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the connection message modal component when connection logs are empty 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="modal"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-xl"
      id="connection-log-modal"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <strong>
            connection.logs
          </strong>
          : 
          mock-connection-id
          <button
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
            <span
              class="sr-only"
            >
              Close
            </span>
          </button>
        </div>
        <div
          class="modal-body"
        >
          <div
            class="text-align-center jumbotron"
            id="empty-logs-jumbotron"
          >
            <p
              class="empty-p"
            >
              info.message.no.logs
            </p>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="uppercase-text btn btn-primary btn-sm"
            id="connection-logs-modal-close-button"
            type="button"
          >
            close
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders the connection message modal component when getting connection logs fails 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    aria-hidden="true"
    class="modal-backdrop show"
  />
  <div
    aria-hidden="true"
    aria-modal="true"
    class="modal"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-xl"
      id="connection-log-modal"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <strong>
            connection.logs
          </strong>
          : 
          mock-connection-id
          <button
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
            <span
              class="sr-only"
            >
              Close
            </span>
          </button>
        </div>
        <div
          class="modal-body"
        >
          <div
            class="text-align-center jumbotron"
            id="empty-logs-jumbotron"
          >
            <p
              class="empty-p"
            >
              info.message.no.logs
            </p>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="uppercase-text btn btn-primary btn-sm"
            id="connection-logs-modal-close-button"
            type="button"
          >
            close
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="modal"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-dialog-centered"
      id="message-modal"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-body"
        >
          <div
            class="fade alert alert-danger show"
            role="alert"
          >
            error.message.get.logs
            <br />
            error
            : 
            <code>
              "Error"
            </code>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="uppercase-text btn btn-primary btn-sm"
            id="message-modal-close-button"
            type="button"
          >
            ok
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders the connection message modal component when there are connection logs 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="modal"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-xl"
      id="connection-log-modal"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <strong>
            connection.logs
          </strong>
          : 
          mock-connection-id
          <button
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
            <span
              class="sr-only"
            >
              Close
            </span>
          </button>
        </div>
        <div
          class="modal-body"
        >
          <table
            class="table table-md"
            id="connection-logs-table"
          >
            <thead>
              <tr>
                <th>
                  message
                </th>
                <th>
                  log.type
                </th>
                <th>
                  date
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <form
                    class=""
                  >
                    <div
                      class="form-group"
                    >
                      <textarea
                        class="form-control"
                        readonly=""
                        rows="5"
                      >
                        {"message":"info message"}
                      </textarea>
                    </div>
                  </form>
                </td>
                <td
                  class="uppercase-text"
                >
                  info
                </td>
                <td>
                  1970-01-01T00:00:00.000Z
                </td>
              </tr>
              <tr>
                <td>
                  <form
                    class=""
                  >
                    <div
                      class="form-group"
                    >
                      <textarea
                        class="form-control"
                        readonly=""
                        rows="5"
                      >
                        {"message":"error message"}
                      </textarea>
                    </div>
                  </form>
                </td>
                <td
                  class="uppercase-text"
                >
                  error
                </td>
                <td>
                  1970-01-01T00:00:00.000Z
                </td>
              </tr>
            </tbody>
          </table>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <button
                class="uppercase-text btn btn-primary btn-sm"
                disabled=""
                id="prev-logs-page-button"
                type="button"
              >
                <i
                  class="bi bi-chevron-double-left"
                />
                 
                prev.page
              </button>
            </div>
            <div
              class="justify-content-end grid col"
            >
              <button
                class="uppercase-text btn btn-primary btn-sm"
                disabled=""
                id="next-logs-page-button"
                type="button"
              >
                next.page
                 
                <i
                  class="bi bi-chevron-double-right"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="uppercase-text btn btn-primary btn-sm"
            id="connection-logs-modal-close-button"
            type="button"
          >
            close
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders the connection message modal component when there are connection logs and the next token 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="modal"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-xl"
      id="connection-log-modal"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <strong>
            connection.logs
          </strong>
          : 
          mock-connection-id
          <button
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
            <span
              class="sr-only"
            >
              Close
            </span>
          </button>
        </div>
        <div
          class="modal-body"
        >
          <table
            class="table table-md"
            id="connection-logs-table"
          >
            <thead>
              <tr>
                <th>
                  message
                </th>
                <th>
                  log.type
                </th>
                <th>
                  date
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <form
                    class=""
                  >
                    <div
                      class="form-group"
                    >
                      <textarea
                        class="form-control"
                        readonly=""
                        rows="5"
                      >
                        {"message":"info message"}
                      </textarea>
                    </div>
                  </form>
                </td>
                <td
                  class="uppercase-text"
                >
                  info
                </td>
                <td>
                  1970-01-01T00:00:00.000Z
                </td>
              </tr>
              <tr>
                <td>
                  <form
                    class=""
                  >
                    <div
                      class="form-group"
                    >
                      <textarea
                        class="form-control"
                        readonly=""
                        rows="5"
                      >
                        {"message":"error message"}
                      </textarea>
                    </div>
                  </form>
                </td>
                <td
                  class="uppercase-text"
                >
                  error
                </td>
                <td>
                  1970-01-01T00:00:00.000Z
                </td>
              </tr>
            </tbody>
          </table>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <button
                class="uppercase-text btn btn-primary btn-sm"
                disabled=""
                id="prev-logs-page-button"
                type="button"
              >
                <i
                  class="bi bi-chevron-double-left"
                />
                 
                prev.page
              </button>
            </div>
            <div
              class="justify-content-end grid col"
            >
              <button
                class="uppercase-text btn btn-primary btn-sm"
                id="next-logs-page-button"
                type="button"
              >
                next.page
                 
                <i
                  class="bi bi-chevron-double-right"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="uppercase-text btn btn-primary btn-sm"
            id="connection-logs-modal-close-button"
            type="button"
          >
            close
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
